include/message_filters/sync_policies/latest_time.h

```diff
@@ -0,0 +1,375 @@
+    }
+  }
+
+private:
+  // assumed data_mutex_ is locked
+  template<int i>
+  void initialize_rate()
+  {
+    if (rate_configs_.size() > 0U) {
+      double rate_ema_alpha{Rate::DEFAULT_RATE_EMA_ALPHA};
+      double error_ema_alpha{Rate::DEFAULT_ERROR_EMA_ALPHA};
+      double rate_step_change_margin_factor{Rate::DEFAULT_MARGIN_FACTOR};
+      if (rate_configs_.size() == RealTypeCount::value) {
+        std::tie (
+          rate_ema_alpha,
+          error_ema_alpha,
+          rate_step_change_margin_factor) = rate_configs_[i];
+      } else if (rate_configs_.size() == 1U) {
+        std::tie (
+          rate_ema_alpha,
+          error_ema_alpha,
+          rate_step_change_margin_factor) = rate_configs_[0U];
+      }
+      rates_.push_back(
+        Rate(
+          ros_clock_->now(),
+          rate_ema_alpha,
+          error_ema_alpha,
+          rate_step_change_margin_factor));
+    } else {
+      rates_.push_back(Rate(ros_clock_->now()));
+    }
+  }
+
+  // assumed data_mutex_ is locked
+  void publish()
+  {
+    parent_->signal(std::get<0>(events_), std::get<1>(events_), std::get<2>(events_),
+                    std::get<3>(events_), std::get<4>(events_), std::get<5>(events_),
+                    std::get<6>(events_), std::get<7>(events_), std::get<8>(events_));
+  }
+
+  struct Rate
+  {
+    static constexpr double DEFAULT_RATE_EMA_ALPHA{0.9};
+    static constexpr double DEFAULT_ERROR_EMA_ALPHA{0.3};
+    static constexpr double DEFAULT_MARGIN_FACTOR{10.0};
+
+    rclcpp::Time prev;
+    double hz{0.0};
+    double error{0.0};
+    double rate_ema_alpha{DEFAULT_RATE_EMA_ALPHA};
+    double error_ema_alpha{DEFAULT_ERROR_EMA_ALPHA};
+    double rate_step_change_margin_factor{DEFAULT_MARGIN_FACTOR};
+    bool do_hz_init{true};
+    bool do_error_init{true};
+
+    Rate(const rclcpp::Time & start)
+      : Rate(start, DEFAULT_RATE_EMA_ALPHA, DEFAULT_ERROR_EMA_ALPHA, DEFAULT_MARGIN_FACTOR)
+    {
+    }
+
+    Rate(const rclcpp::Time & start,
+      const double & rate_ema_alpha, const double & error_ema_alpha,
+      const double & rate_step_change_margin_factor)
+      : prev{start},
+        rate_ema_alpha{rate_ema_alpha},
+        error_ema_alpha{error_ema_alpha},
+        rate_step_change_margin_factor{rate_step_change_margin_factor}
+    {
+      std::cout << "prev{" << prev.seconds() << 
+        "}, rate_ema_alpha{" << rate_ema_alpha << "}, error_ema_alpha{" << error_ema_alpha <<
+        "}, rate_step_change_margin_factor{" << rate_step_change_margin_factor << "}" << std::endl;
+    }
+
+    bool operator>(const Rate & that) const
+    {
+      return this->hz > that.hz;
+    }
+
+    void compute_hz(const rclcpp::Time & now)
+    {
+      bool step_change_detected = false;
+      do {
+        double period = (now-prev).seconds();
+        RCUTILS_ASSERT(period > 0.0 && "Multiple messages received and time is not updating.");
```

Comment:
I'm not a huge fan of this.  This resolves into an `assert` statement, which means that when building in release mode, this protection won't be in place.  I don't think this is a very performance-sensitive piece of code, so I'd rather that we just make this an if statement that exits the function if we see it.

Reply:
done